### PR TITLE
Better parse when lambda expressions contain default parameter values.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/58644

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@RikkiGibson this is ready for review.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -13107,10 +13111,23 @@ private ParameterSyntax ParseLambdaParameter()
             var identifier = this.ParseIdentifierToken();
             ParseParameterNullCheck(out var exclamationExclamationToken, out var equalsToken);
 
-            // 'equalsToken' should always be null here. Otherwise, it means while we were scanning
-            // for potential lambdas, we saw an `!!=` or `!=` in the parameter list and allowed
-            // it. However, that should never happen as ScanParenthesizedLambda allows `!!` only.
-            Debug.Assert(equalsToken == null);
+            // If we didn't already consume an equals sign as part of !!=, then try to scan one out now. Note: this is
+            // not legal code.  But we detect it so that we can give the user a good message, and so we don't completely
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:13115]: "don't *go* completely"?

**Review by RikkiGibson — APPROVED** [review]
That was easy!

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler for a second review on a relatively simple change

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler for small parser change that produces better error recovery results.  thanks!

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -13107,10 +13111,23 @@ private ParameterSyntax ParseLambdaParameter()
...
-            // it. However, that should never happen as ScanParenthesizedLambda allows `!!` only.
-            Debug.Assert(equalsToken == null);
+            // If we didn't already consume an equals sign as part of !!=, then try to scan one out now. Note: this is
+            // not legal code.  But we detect it so that we can give the user a good message, and so we don't go
+            // completely off the rails.
+            //
+            // Note: we add the `= value` as skipped trivia to either the identifier or `!!` (if we have the latter).
+            // This allows us to handle this code without ever showing it the binding phases.  We could consider
+            // actually binding this in the future if it would be helpful and if we're ok paying the testing cost of
+            // checking this at the semantic layers.
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:13121]: Consider dropping the end of the comment - we typically bind code the user wrote, even in error situations, so the cost shouldn't be an issue.
```suggestion
            // This allows us to handle this code without ever showing it to the binding phases.  We could consider
            // actually binding this in the future.
```

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -4415,6 +4420,27 @@ private ParameterSyntax ParseParameter()
                 if (equalsToken != null)
                     equalsValueClause = CheckFeatureAvailability(_syntaxFactory.EqualsValueClause(equalsToken, this.ParseExpressionCore()), MessageID.IDS_FeatureOptionalParameter);
 
+                if (type == null)
+                {
+                    // in the __arglist case, all other parameter pieces are not legal.  So attach anything else we get
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4425]:
```suggestion
                    // In the __arglist case, none of the other parameter pieces are legal, so attach anything else we get
```

**Review by 333fred — APPROVED** [review]

**Comment by cston** [comment]
>                         attributes, modifiers.ToList(), type: null, this.EatToken(SyntaxKind.ArgListKeyword), exclamationExclamationToken: null, @default: null);

I'm not sure it's worth handling `__arglist!!` and `__arglist = value`. The resulting code seems more complicated for a very unlikely scenario.

---
Refers to: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:4395 in 0613817. [](commit_id = 0613817bdba10db04ec072d680d17aa8137ef541, deletion_comment = True)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:13147]: > so the cost shouldn't be an issue.

In this case, i mentioned that it's "testing cost" :)  In other words, i'd like to have the better error recovery here, but i don't really want to write the tons of semantic tests that validate this :)

**Comment by CyrusNajmabadi** [comment]
> I'm not sure it's worth handling __arglist!! and __arglist = value. The resulting code seems more complicated for a very unlikely scenario.

It didn't bother me to support.  I'm a fan of the parser not going off the rails for sensible constructs.  I do agree it's exceedingly unlikely for this code to happen though (who even uses __arglist), so i can revert this though if you'd like. :)

**Review by cston — APPROVED** [review]

**Review by ryzngard — APPROVED** [review]
Test update lgtm

**Pull request merged**